### PR TITLE
JS: add simple query for detecting sensitive files downloaded over insecure connection

### PR DESCRIPTION
**Pull request by erik-krogh**
CVE-2017-16003: TP/TN

A very closely related issue to look for `http` urls in 'package.json` files, or `<script />` tags with `http` urls, but that is not a path-problem, so it feels weird mashing that into the same query.   
Maybe we should get back to that later? 

It was suggested at one point that I could add an integrity-check as a sanitizer, but I don't know what that would look like?

The query only supports the `http` and `ftp` protocols for now. Suggestions for additional protocols are very welcome. 

Performance evaluation waits until after this PR has been merged into the sprint branch.

### CODE REVIEW

File: javascript/ql/src/Security/CWE-829/UnsecureDownload.ql [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// An integrity-check is a sanitizer (but what does such a check look like?)
+import javascript
+import DataFlow::PathGraph
+
+class Configuration extends DataFlow::Configuration {
+  Configuration() { this = "HTTP/HTTPS" }
+
+  override predicate isSource(DataFlow::Node source) {
+    exists(string str | str = source.getStringValue() |
+      str.regexpMatch("http://.*|ftp://.'") and
```
Comment (asgerf) [inline javascript/ql/src/Security/CWE-829/UnsecureDownload.ql:26]:
```suggestion
      str.regexpMatch("http://.*|ftp://.*") and
```

File: javascript/ql/src/Security/CWE-829/UnsecureDownload.ql [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/**
+ * @name Download of sensitive file through unsecure connection
+ * @description Downloading executeables and other sensitive files over an unsecure connection
```
Comment (asgerf) [inline javascript/ql/src/Security/CWE-829/UnsecureDownload.ql:3]:
```suggestion
 * @description Downloading executables and other sensitive files over an insecure connection
```

File: javascript/ql/src/semmle/javascript/security/dataflow/InsecureDownloadCustomizations.qll [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /**
+   * Gets a file-extension that can potentially be dangerous.
+   *
+   * Archives are included, because they often contain source-code.
+   */
+  string unsafeExtension() {
+    result =
+      ["exe", "dmg", "pkg", "tar.gz", "zip", "sh", "bat", "cmd", "app", "apk", "msi", "dmg",
```
Comment (intrigus-lgtm) [inline javascript/ql/src/semmle/javascript/security/dataflow/InsecureDownloadCustomizations.qll:55]: What about `.jar` and `.war` files?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/InsecureDownloadCustomizations.qll:55]: Good point, those after often used as executeables.

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+			Downloading executeables or other sensitive files over an unencrypted connection 
+			can leave a server open to man-in-the-middle attacks (MITM). 
+			Such a man-in-the-middle attack can allow an attacker to insert arbitary content 
+			into the downloaded file, and in the worst case allow the attacker to execute 
+			arbitary code on the vulnerable system.
+		</p>
+	</overview>
+	<recommendation>
+		<p>
+			Use an transfer protocol that includes encryption when downloading executeables or other sensitive files.
```
Comment (asgerf) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:16]:
```suggestion
			Use a secure transfer protocol when downloading executeables or other sensitive files.
```

File: javascript/ql/src/semmle/javascript/security/dataflow/InsecureDownload.qll [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * Classes and predicates for reasoning about download of sensitive file through insecure connection vulnerabilities.
+ */
+module InsecureDownload {
+  import InsecureDownloadCustomizations::InsecureDownload
+
+  /**
+   * A taint tracking configuration for download of sensitive file through insecure connection.
+   */
+  class Configuration extends DataFlow::Configuration {
+    Configuration() { this = "HTTP/HTTPS" }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/InsecureDownload.qll:21]:
```suggestion
    Configuration() { this = "InsecureDownload" }
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+	<overview>
+		<p>
+			Downloading executeables or other sensitive files over an unencrypted connection 
+			can leave a server open to man-in-the-middle attacks (MITM). 
+			Such a man-in-the-middle attack can allow an attacker to insert arbitary content
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:9]:
```suggestion
			Such an attack can allow an attacker to insert arbitrary content 
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 7]
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+	<overview>
+		<p>
+			Downloading executeables or other sensitive files over an unencrypted connection 
+			can leave a server open to man-in-the-middle attacks (MITM). 
+			Such a man-in-the-middle attack can allow an attacker to insert arbitary content 
+			into the downloaded file, and in the worst case allow the attacker to execute
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:10]:
```suggestion
			into the downloaded file, and in the worst case, allow the attacker to execute 
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+	<overview>
+		<p>
+			Downloading executeables or other sensitive files over an unencrypted connection 
+			can leave a server open to man-in-the-middle attacks (MITM). 
+			Such a man-in-the-middle attack can allow an attacker to insert arbitary content 
+			into the downloaded file, and in the worst case allow the attacker to execute 
+			arbitary code on the vulnerable system.
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:11]:
```suggestion
			arbitrary code on the vulnerable system.
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+		</p>
+	</overview>
+	<recommendation>
+		<p>
+			Use a secure transfer protocol when downloading executeables or other sensitive files.
+		</p>
+	</recommendation>
+	<example>
+		<p>
+			In this example a server downloads a shell script from a remote URL using the <code>node-fetch</code>
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:21]:
```suggestion
			In this example, a server downloads a shell script from a remote URL using the <code>node-fetch</code> 
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+			Downloading executeables or other sensitive files over an unencrypted connection 
+			can leave a server open to man-in-the-middle attacks (MITM). 
+			Such a man-in-the-middle attack can allow an attacker to insert arbitary content 
+			into the downloaded file, and in the worst case allow the attacker to execute 
+			arbitary code on the vulnerable system.
+		</p>
+	</overview>
+	<recommendation>
+		<p>
+			Use a secure transfer protocol when downloading executeables or other sensitive files.
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:16]:
```suggestion
			Use a secure transfer protocol when downloading executables or other sensitive files.
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+	</recommendation>
+	<example>
+		<p>
+			In this example a server downloads a shell script from a remote URL using the <code>node-fetch</code> 
+			library, and then executes this shell script. 
+		</p>
+		<sample src="examples/insecure-download.js" />
+		<p>
+			The HTTP protocol is vulnerable to MITM, and thus an attacker could potentially replace the downloaded
+			shell script with arbitary code, which allows the attacker complete control over the attacked system.
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:27]:
```suggestion
			shell script with arbitrary code, which gives the attacker complete control over the system.
```

File: javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+		</p>
+		<p>
+			The issue has been fixed in the below example by replacing the HTTP protocol with the HTTPS protocol.
```
Comment (mchammer01) [inline javascript/ql/src/Security/CWE-829/InsecureDownload.qhelp:30]:
```suggestion
			The issue has been fixed in the example below by replacing the HTTP protocol with the HTTPS protocol.
```

**Comment by intrigus-lgtm** [comment]
I'm not sure how relevant this is in practice:
When I looked at a similar query for java, I found instances where the download link looked like this:
`http://foo.foo/latestVersion` (Notice: No extension). This would then be saved to a `version.exe` file.

So it **may** be useful to also to track such cases. Or maybe not :man_shrugging:

**Comment by erik-krogh** [comment]
> I'm not sure how relevant this is in practice:
> When I looked at a similar query for java, I found instances where the download link looked like this:
> `http://foo.foo/latestVersion` (Notice: No extension). This would then be saved to a `version.exe` file.
> 
> So it **may** be useful to also to track such cases. Or maybe not

I think it is useful to track such cases. 

And you got a good point, thanks :+1: 
But it is a big change that is better suited for a later PR.

**Comment by erik-krogh** [comment]
> Happy for this to be merged once you had a look at my suggestions and comments (there are a few typos, but apart from that, it's looking really good).
> Hope it helps!

It did, I agree with all your suggested changes :+1:

**Pull request merged**